lib/events.js

```diff
@@ -89,8 +113,47 @@ EventEmitter.init = function() {
   }
 
   this._maxListeners = this._maxListeners || undefined;
+
+  if (opts && opts.captureRejections) {
+    this[kCapture] = !!opts.captureRejections;
+  } else {
+    // Assigning it directly a prototype lookup, as it slighly expensive
+    // and it sits in a very sensitive hot path.
+    this[kCapture] = EventEmitter.prototype[kCapture];
+  }
 };
 
+function addCatch(that, promise, type, args) {
+  // Handle Promises/A+ spec, then could be a getter
+  // that throws on second use.
+  try {
+    const then = promise.then;
+
+    if (typeof then === 'function') {
+      then.call(promise, undefined, (err) => {
+        // The callback is called with nextTick to avoid a follow-up
+        // rejection from this promise.
+        process.nextTick(emitUnhandledRejectionOrErr, that, err, type, args);
+      });
+    }
+  } catch (err) {
+    that.emit('error', err);
+  }
+}
+
+function emitUnhandledRejectionOrErr(ee, err, type, args) {
+  if (typeof ee[kRejection] === 'function') {
+    ee[kRejection](err, type, ...args);
+  } else {
+    // We have disable the capture rejections mechanism, otherwise
+    // we might end up in an infinite loop.
+    const prev = ee[kCapture];
+    ee[kCapture] = false;
+    ee.emit('error', err);
```

Comment:
nit - this setting of ee[kCapture] should ideally be in a try/finally to handle the case the error handler throws

Reply:
This turned out to be a significant change in how the machinery work in this case, and it was definitely an edge case that had to be handled. Good spot!